clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```
@@ -365,11 +369,20 @@ private static int configureRequestTimeout(ProducerConfig config, Map<String, Ob
 
     private static TransactionState configureTransactionState(ProducerConfig config, Time time) {
         boolean idempotenceEnabled = config.getBoolean(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG);
+        TransactionState transactionState = null;
         if (idempotenceEnabled) {
-            return new TransactionState(time);
-        } else {
-            return null;
+            String transactionalId = config.getString(ProducerConfig.TRANSACTIONAL_ID_CONFIG);
+            int transactionTimeoutMs = config.getInt(ProducerConfig.TRANSACTION_TIMEOUT_CONFIG);
+            transactionState = new TransactionState(transactionalId, transactionTimeoutMs);
+            if (transactionState.isTransactional())
+                log.info("Instantiated a transactional producer.");
+            else
+                log.info("Instantiated an idempotent producer.");
+        } else if (config.originals().containsKey(ProducerConfig.TRANSACTIONAL_ID_CONFIG)) {
```

Perhaps if the user configures a transactionalId, then we should enable idempotence automatically. We can raise an exception only if the user has explicitly disabled idempotence.

---

One concern is making things too magical. However, it does seem that there's no scenario where someone would want transactions without idempotence. So, I'm in favour of this suggestion.